src/JustSaying/Messaging/Channels/Multiplexer/RoundRobinQueueMultiplexer.cs

```diff
@@ -0,0 +1,179 @@
+
+        private readonly SemaphoreSlim _readersLock = new SemaphoreSlim(1, 1);
+        private readonly object _startLock = new object();
+
+        private bool _started = false;
+        private CancellationToken _stoppingToken;
+        private Task _completion;
+
+        public RoundRobinQueueMultiplexer(
+            int channelCapacity,
+            ILogger<RoundRobinQueueMultiplexer> logger)
+        {
+            _readers = new List<ChannelReader<IQueueMessageContext>>();
+            _logger = logger;
+
+            _channelCapacity = channelCapacity;
+            _targetChannel = Channel.CreateBounded<IQueueMessageContext>(_channelCapacity);
+        }
+
+        public void ReadFrom(ChannelReader<IQueueMessageContext> reader)
+        {
+            if (reader == null) throw new ArgumentNullException(nameof(reader));
+
+            _readersLock.Wait(_stoppingToken);
+
+            try
+            {
+                _readers.Add(reader);
+            }
+            finally
+            {
+                _readersLock.Release();
+            }
+
+            async Task OnReaderCompletion()
+            {
+                await reader.Completion.ConfigureAwait(false);
+                RemoveReader(reader);
+            }
+
+            _ = OnReaderCompletion();
+        }
+
+        private void RemoveReader(ChannelReader<IQueueMessageContext> reader)
+        {
+            _logger.LogInformation("Received notification to remove reader from multiplexer inputs");
+
+            _readersLock.Wait(_stoppingToken);
+
+            try
+            {
+                _readers.Remove(reader);
+            }
+            finally
+            {
+                _readersLock.Release();
+            }
+        }
+
+        public Task Run(CancellationToken stoppingToken)
+        {
+
+            if (!_started)
+            {
+                lock (_startLock)
+                {
+                    if (!_started)
+                    {
+                        _stoppingToken = stoppingToken;
+                        _completion = RunImpl();
+                        _started = true;
+                    }
+                }
+            }
+
+            return _completion;
+        }
+
+        private async Task RunImpl()
+        {
+            await Task.Yield();
+
+            _logger.LogInformation(
+                "Starting up channel multiplexer with a queue capacity of {Capacity}",
+                _channelCapacity);
+
+            ChannelWriter<IQueueMessageContext> writer = _targetChannel.Writer;
+            while (true)
+            {
+                await _readersLock.WaitAsync(_stoppingToken).ConfigureAwait(false);
+
+                _stoppingToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    if (_readers.Count < 1)
+                    {
+                        _logger.LogInformation("All writers have completed, terminating multiplexer");
```

Comment:
Should this be "all readers have completed"?

Reply:
Internally they're 'readers' to the multiplexer, but at a system level they are 'writers' into the multiplexer, so it makes sense for someone reading the log message.

I don't think we need this check at all anyway now, as the only reason readers are cancelled is when the bus is cancelled, and that will tear this down anyway (i.e. we're not really using channel completion here).

Reply:
No longer an issue as this has been superseded by the new MergingMultiplexer